Notification overhaul

kurtwheeler opened the pull request:
## Issue Number

#312 

## Purpose/Implementation Notes

This PR does a lot. It completely replaces all notification types with new ones. It changes the way notifications are templated and sent. It changes how notifications are generated in order to send them to entire teams. It changes every test that tests notifications.

I am sorry for how big this got, but there wasn't really any midpoint I could have stopped at without things being broken.

## Types of changes

- Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Functional tests

I've got a lot of unit tests. I've looked at the email configs that get built, all of them. They look fairly good, but really I think this PR 
will need a lot of staging testing to make sure all the formatting works well. Can't really see it till you get one.

## Checklist

- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)
- [x] Any dependent changes have been merged and published in downstream modules

kurtwheeler commented:
@Philinphiladelphia I think this PR is finally ready for review!!!

@dvenprasad it might be good for you to look through the notifications_config.

I am really sorry about how huge this PR got, but it wasn't until just now that things stopped being broken.

I'm not sure that every notification that needs to be sent will be, but I'd rather address that in subsequent PRs. I'll look through tomorrow and see if I can spot any to make issues for.

kurtwheeler commented:
https://github.com/AlexsLemonade/resources-portal/issues/371 Has the notifications that I still need to generate and/or test for. I will do these in a subsequent PR once this is merged. (It shouldn't be too hard to knock all those out, the code to make a notification is pretty minimal at this point.)

Philinphiladelphia commented on api/resources_portal/emailer.py:
```diff
@@ -7,6 +7,38 @@
+    if settings.AWS_SES_DOMAIN
+    else "Resources Portal Mail Robot"
+)
+EMAIL_SOURCE_TEMPLATE = "Resources Portal Mail Robot <no-reply@{}>"
```
It might make sense to use this template in line 11 above, just so we have one unified source for the "Resources Portal Mail Robot" text.

Philinphiladelphia commented on api/resources_portal/migrations/0026_auto_20200901_1642.py:
```diff
@@ -0,0 +1,148 @@
+# Generated by Django 2.2.13 on 2020-09-01 16:42
```
We're already up to 0026 on dev. This migration will need to be remade.

Philinphiladelphia commented on api/resources_portal/test/models/test_notifications_config.py:
```diff
@@ -0,0 +1,72 @@
+            self.notification.save()
+            email_dict = self.notification.get_email_dict()
+
+            # We formtat the html different than the other fields.
```
```suggestion
            # We format the html different than the other fields.
```

Philinphiladelphia commented on api/resources_portal/test/views/test_material_request_issue.py:
```diff
@@ -198,30 +197,26 @@ def test_put_request_from_sharer_updates_a_material_request(self):
-            ),
-            1,
-        )
+        # Do we need an equivalent notification for this?
```
Probably not? It's a pretty minute action.

dvenprasad commented on api/resources_portal/models/notifications_config.py:
```diff
@@ -0,0 +1,539 @@
+    },
+    "MATERIAL_REQUEST_SHARER_CANCELLED": {
+        "subject": "Request for {material_category} cancelled",
+        "body": "{requester_name} rejected a request for {material_category}, {material_name}.",
```
```suggestion
        "body": "{requester_name} cancelled a request for {material_category}, {material_name}.",
```

dvenprasad commented on api/resources_portal/models/notifications_config.py:
```diff
@@ -0,0 +1,539 @@
+        "CTA_link_field": "associated_material_request",
+        "plain_text_email": (
+            "{your_name},\n{requester_name} provided the following required items for a request for"
+            " {material_category}, {material_name}:{provided_info_plain}\n\nReview items ({request_url})."
```
Can {provided_info_plain} be on a new line?

dvenprasad commented on api/resources_portal/models/notifications_config.py:
```diff
@@ -0,0 +1,539 @@
+            "{organization_name} marked your request for {material_category},"
+            " {material_name} as fulfilled.\nPlease view fulfilment notes for details."
+        ),
+        "CTA": "View Request",
```
```suggestion
        "CTA": "View Fulfillment Notes",
```

dvenprasad reviewed (changes requested):
Most of it looks good. I've left some suggestions.

Also I noticed that these notifications were missing. 
- Sharer accepts a request and the request does not require any additional info from the requester 
- Escalating to Grants team

Is that being handled elsewhere?

kurtwheeler commented:
> Also I noticed that these notifications were missing. 
> - Sharer accepts a request and the request does not require any additional info from the requester 
> - Escalating to Grants team

> Is that being handled elsewhere?

Yes. The first should just be the IN_FULFILLMENT notification because if the request does not need any additional info from the requester it can skip the APPROVED status. The second has it's own PR/issue: https://github.com/AlexsLemonade/resources-portal/pull/373 and https://github.com/AlexsLemonade/resources-portal/issues/374

dvenprasad commented:
> The first should just be the IN_FULFILLMENT notification because if the request does not need any additional info from the requester

Yes, the emails to the requesters are good.
I think the sharer email for this scenario is not there, saying that the request was accepted.  i think it is this one:
```
Subject: Request for <resource type> accepted
Body: You/<Team Member> accepted a request for <resource type>, <resource name> from <Requester>.
CTA: View Request

Plain Text
<Full Name>,
You/<Team Member> accepted a request for <resource type>, <resource name> from <Requester>.

View request details.

The CCRR Team
-----
You are receiving this email because you subscribed to receive notifications from CCRR portal. Manage Notifications
Alex's Lemonade Stand Foundation
111 Presidential Blvd, Suite 203, Bala Cynwyd,  PA 19004
```

kurtwheeler commented:
Ah okay, I think I gotcha. Basically we need a notification to fire if it goes straight to IN_FULFILLMENT. The MATERIAL_REQUEST_SHARER_EXECUTED_MTA` notification says:`Please make arrangements to send the {material_category} to {requester_name}."` Should we include something like that in this one too?

kurtwheeler commented:
OK @Philinphiladelphia I think I addressed all your feedback (I've got https://github.com/AlexsLemonade/resources-portal/issues/372 to track the cron job.)

@dvenprasad I think I've now addressed your concerns as well, although I think maybe we should tweak the copy a tad for the notification I just added?

dvenprasad commented:
Okay, yeah let's add that sentence in. Good call 🚀

Philinphiladelphia reviewed (commented):
Looks like everything I mentioned got wrapped up. Merge when ready!
